### PR TITLE
WIP: Reviews Encouraged - Eject gogo from operator

### PR DESCRIPTION
**Pull request by sdake**
The operator appears to have a nonstandard CRD interface. This PR
does two things. It standardizes (partially) on the upstream K8s
CRD interface for `IstioControlPlane` and additionally removes
most of gogo. As gogo is a performance tuning system, there is no
strong motivation to use gogo. The usage of gogo is trigging confusing
type conversions in the code base including marshal to translate
values. Further because of bugs in gogo, some funky translation of
go structs and sedding of output variables is happening - which is wrong :)

`Co-Authored-By: Morven Cao <llcao@cn.ibm.com>`

Thanks to @morvencaoMorven for triggering the thinking process
after root causing with his work in PR: https://github.com/istio/operator/pull/381

### CODE REVIEW

File: deploy/operator.yaml [hunk 1]
```diff
@@ -16,7 +16,7 @@ spec:
       serviceAccountName: istio-operator
       containers:
         - name: istio-operator
-          image: gcr.io/istio-testing/operator:master-latest-daily
+          image: docker.io/sdake/operator:sdake
```
Comment (sdake) [inline deploy/operator.yaml]: note to change this to your own docker hub if testing manually - or alternately revert this part of the PR when the work is done.

File: Makefile.core.mk [hunk 2]
```diff
@@ -118,8 +118,8 @@ types_v1alpha2_openapi := $(types_v1alpha2_protos:.proto=.json)
 $(types_v1alpha2_pb_gos) $(types_v1alpha2_pb_docs) $(types_v1alpha2_pb_pythons): $(types_v1alpha2_protos)
 	@$(protoc) $(go_plugin) $(protoc_gen_k8s_support_plugins) $(protoc_gen_docs_plugin)$(types_v1alpha2_path) $(protoc_gen_python_plugin) $^
 	@cp -r ${TMPDIR}/pkg/* pkg/
-	@sed -i 's|github.com/gogo/protobuf/protobuf/google/protobuf|github.com/gogo/protobuf/types|g' $(types_v1alpha2_path)/istiocontrolplane_types.pb.go
-	@go run $(repo_dir)/pkg/apis/istio/fixup_structs/main.go -f $(types_v1alpha2_path)/istiocontrolplane_types.pb.go
+#	@sed -i 's|github.com/gogo/protobuf/protobuf/google/protobuf|github.com/gogo/protobuf/types|g' $(types_v1alpha2_path)/istiocontrolplane_types.pb.go
```
Comment (sdake) [inline Makefile.core.mk]: remove dead code.

File: Makefile.core.mk [hunk 3]
```diff
@@ -138,8 +138,8 @@ values_v1alpha1_openapi := $(values_v1alpha1_protos:.proto=.json)
 $(values_v1alpha1_pb_gos) $(values_v1alpha1_pb_docs) $(values_v1alpha1_pb_pythons): $(values_v1alpha1_protos)
 	@$(protoc) $(go_plugin) $(protoc_gen_k8s_support_plugins) $(protoc_gen_docs_plugin)$(values_v1alpha1_path) $(protoc_gen_python_plugin) $^
 	@cp -r ${TMPDIR}/pkg/* pkg/
-	@sed -i 's|github.com/gogo/protobuf/protobuf/google/protobuf|github.com/gogo/protobuf/types|g' $(values_v1alpha1_path)/values_types.pb.go
-	@go run $(repo_dir)/pkg/apis/istio/fixup_structs/main.go -f $(values_v1alpha1_path)/values_types.pb.go
+#	@sed -i 's|github.com/gogo/protobuf/protobuf/google/protobuf|github.com/gogo/protobuf/types|g' $(values_v1alpha1_path)/values_types.pb.go
```
Comment (sdake) [inline Makefile.core.mk]: remove dead code.

File: go.mod [hunk 4]
```diff
@@ -20,7 +20,7 @@ require (
 	github.com/go-logr/logr v0.1.0
 	github.com/go-logr/zapr v0.1.1 // indirect
 	github.com/gobwas/glob v0.2.3 // indirect
-	github.com/gogo/protobuf v1.3.0
+	github.com/gogo/protobuf v1.2.1
```
Comment (sdake) [inline go.mod]: gogo should not be showing up in gomod - especially 1.2.1.

File: pkg/apis/istio/v1alpha2/deepcopy.go [hunk 5]
```diff
@@ -15,7 +15,7 @@
 package v1alpha2
 
 import (
-	"github.com/gogo/protobuf/proto"
+       "github.com/gogo/protobuf/proto"
```
Comment (sdake) [inline pkg/apis/istio/v1alpha2/deepcopy.go]: need to find the equivalent repo without gogo - any hints would be helpful 💯

File: pkg/apis/istio/v1alpha2/istiocontrolplane_types.proto [hunk 6]
```diff
@@ -166,12 +166,12 @@ package v1alpha2;
 
 // IstioControlPlane is a CustomResourceDefinition (CRD) describing an Istio control plane.
 message IstioControlPlane {
+    // Defines the object meta
+    k8s.io.apimachinery.pkg.apis.meta.v1.ObjectMeta metadata = 1;
```
Comment (sdake) [inline pkg/apis/istio/v1alpha2/istiocontrolplane_types.proto]: this is correct, but there should be another field added to `IstioControlPlane`.  Unfortunately I am not quite sure what it is - TypeMeta iirc...

File: pkg/controller/istiocontrolplane/istiocontrolplane_controller.go [hunk 7]
```diff
@@ -128,17 +128,17 @@ func (r *ReconcileIstioControlPlane) Reconcile(request reconcile.Request) (recon
 		}
 		// TODO: for now, nuke the resources, regardless of errors
```
Comment (sdake) [inline pkg/controller/istiocontrolplane/istiocontrolplane_controller.go]: remove the word nuke.

File: pkg/name/name.go [hunk 8]
```diff
@@ -140,9 +140,9 @@ func IsFeatureEnabledInSpec(featureName FeatureName, controlPlaneSpec *v1alpha2.
 // TODO: remove extra validations when comfort level is high enough.
 func IsComponentEnabledInSpec(featureName FeatureName, componentName ComponentName, controlPlaneSpec *v1alpha2.IstioControlPlaneSpec) (bool, error) {
 	//check in Values part as well for third Party components
-	if featureName == ThirdPartyFeatureName {
-		return IsComponentEnabledFromValue(string(componentName), controlPlaneSpec.Values)
-	}
+//	if featureName == ThirdPartyFeatureName {
```
Comment (sdake) [inline pkg/name/name.go]: not sure why this is commented out - uncomment and test.

File: pkg/util/yaml.go [hunk 9]
```diff
@@ -18,9 +18,11 @@ import (
 	"bytes"
 	"strings"
 
-	"github.com/gogo/protobuf/jsonpb"
-	"github.com/gogo/protobuf/proto"
+       "github.com/gogo/protobuf/proto"
```
Comment (sdake) [inline pkg/util/yaml.go]: I think this is wrong - and the logic to convert one json to another may really just be overhead that is unneeded.

**Review by sdake — COMMENTED** [review]
fails when installing CR with
```
sdake@beast-01:~/go/src/istio.io/operator$ kubectl logs -n istio-operator istio-operator-5c94bb8c94-t2t6x
2019-10-15T16:05:01.183900Z	info	ControlZ available at 127.0.0.1:9876
2019-10-15T16:05:01.489547Z	info	Registering Components.
2019-10-15T16:05:01.489658Z	info	Adding controller for IstioControlPlane
2019-10-15T16:05:01.491089Z	info	Controller added
2019-10-15T16:05:01.491141Z	info	Starting the Cmd.
2019-10-15T16:05:01.491272Z	info	attempting to acquire leader lease  istio-operator/istio-operator-lock...
2019-10-15T16:05:01.510144Z	info	successfully acquired lease istio-operator/istio-operator-lock
2019-10-15T16:05:46.040865Z	error	pkg/mod/k8s.io/client-go@v0.0.0-20190620085101-78d2af792bab/tools/cache/reflector.go:98: unable to understand watch event watch.Event{Type:"ADDED", Object:(*v1alpha2.IstioControlPlane)(0xc0007e8c40)}
2019-10-15T16:11:17.522764Z	error	pkg/mod/k8s.io/client-go@v0.0.0-20190620085101-78d2af792bab/tools/cache/reflector.go:98: unable to understand watch event watch.Event{Type:"ADDED", Object:(*v1alpha2.IstioControlPlane)(0xc0007e87c0)}
2019-10-15T16:20:19.524409Z	error	pkg/mod/k8s.io/client-go@v0.0.0-20190620085101-78d2af792bab/tools/cache/reflector.go:98: unable to understand watch event watch.Event{Type:"ADDED", Object:(*v1alpha2.IstioControlPlane)(0xc000691d00)}
2019-10-15T16:26:03.526101Z	error	pkg/mod/k8s.io/client-go@v0.0.0-20190620085101-78d2af792bab/tools/cache/reflector.go:98: unable to understand watch event watch.Event{Type:"ADDED", Object:(*v1alpha2.IstioControlPlane)(0xc0007e8640)}
```

**Comment by sdake** [comment]
This looks far superior in terms of output - even though things don't appear to be working just yet ```sdake@beast-01:~/go/src/istio.io/operator$ kubectl get istiocontrolplane -n istio-operator example-istiocontrolplane -oyaml
apiVersion: install.istio.io/v1alpha2
kind: IstioControlPlane
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"install.istio.io/v1alpha2","kind":"IstioControlPlane","metadata":{"annotations":{},"name":"example-istiocontrolplane","namespace":"istio-operator"},"spec":{"profile":"default"}}
  creationTimestamp: "2019-10-15T16:05:46Z"
  generation: 1
  name: example-istiocontrolplane
  namespace: istio-operator
  resourceVersion: "864"
  selfLink: /apis/install.istio.io/v1alpha2/namespaces/istio-operator/istiocontrolplanes/example-istiocontrolplane
  uid: a5962770-ef65-11e9-8369-0042688109a5
spec:
  profile: default
```

**Comment by sdake** [comment]
Deletion works correctly within milliseconds - although Istio wasn't actually deployed because of above error.
```
sdake@beast-01:~/go/src/istio.io/operator$ kubectl delete istiocontrolplane example-istiocontrolplane -n istio-operator
istiocontrolplane.install.istio.io "example-istiocontrolplane" deleted
sdake@beast-01:~/go/src/istio.io/operator$ kubectl get istiocontrolplanes -n istio-operator
No resources found.
```

File: pkg/apis/istio/v1alpha2/istiocontrolplane_types.proto (hunk 6 above)
Comment (ostromart) [reply pkg/apis/istio/v1alpha2/istiocontrolplane_types.proto]: This approach is not going to work, unless we abandon proto as a source of truth (which we discussed with @geeknoid in the past). There's no way to create an anonymous field in go structs from proto. 
Please see https://github.com/istio/operator/pull/370 for workaround.

**Comment by geeknoid** [comment]
Note that this PR will have some conflicts with https://github.com/istio/operator/pull/370 I believe.

File: pkg/apis/istio/v1alpha2/istiocontrolplane_types.proto (hunk 6 above)
Comment (sdake) [reply pkg/apis/istio/v1alpha2/istiocontrolplane_types.proto]: I’ll include #370 in this PR and get it in working with order post KUBECON recovery.

Comment (sdake) [reply pkg/apis/istio/v1alpha2/istiocontrolplane_types.proto]: I’ll include #370 in this PR and get it in working with order post KUBECON recovery.

**Comment by sdake** [comment]
Closing PR as this will need rework in master of istio/istio.

**Pull request closed without merging**